### ISSUE TITLE
Stubbing Multiple Consecutive Results

### ISSUE THREAD

**Issue by dotnetspec** [brian-watkins/elm-spec#72]
Hi,

Thanks for creating the package. It looks very promising.

I've set it up in my project and successfully imported and used many of the modules e.g.

```
import Spec.Markup as Markup
import Spec.Markup.Event as Event
import Spec.Markup.Selector exposing (by, id, tag)
import Spec.Report as Report exposing (Report)
import Spec.Setup exposing (Setup)
import Spec.Step as Step exposing (Step)
import Url exposing (Protocol(..), Url)
```

to get a minimal spec-test in the terminal.

For context I'm testing my app at:
https://squashpassion.com/schedule
(production version)

I'm trying to do something like:
```
displayAvailableTimeslots : Spec App.Model App.Msg
displayAvailableTimeslots =
    let
        testFlag : String
        testFlag =
            "http://localhost:5500/schedule"

        -- "https://squashpassion.com:443"
    in
    describe "Display Available Timeslots"
        [ scenario "User loads the schedule page"
            (given
                --"the user loads the schedule page"
                -- Set up the initial conditions for the test
                (MySpecSetup.initAppSetup
                    testFlag
                    Consts.placeholderURL
                )
                |> when "the page loads"
                    -- Simulate the page loading, which may involve HTTP requests or other actions
                    [ **Spec.Step Step.model (App.GotScheduleMsg Mocks.mockHttpResponse)**
                    ]
                |> it "confirms page receives available timeslots"
                    (Markup.observeElement
                        |> Markup.query
                        << by [ tag "h1" ]
                        |> Spec.expect
                            (Claim.isSomethingWhere <|
                                Markup.text <|
                                    Claim.isStringContaining 1 "Availability Today"
                            )
                    )
            )
        ]
```
        
        
** is line where I'm struggling. I'm not sure if I should be using a Context or 'Step' type alias or something else here. Either way I'm getting:
 
 I cannot find a `Step` variant:

105|                       Step Step.model (App.GotScheduleMsg Mocks.mockHttpResponse)

I'm still implementing the testing locally. I only want to test, in this project at least, that either the timeslots appear or a message stating that there are no more timeslots available today.

When I attempt:
`import Spec.Step.Context exposing (..)`

I get:
```
You are trying to import a `Spec.Step.Context` module:

33| import Spec.Step.Context exposing (..)
           ^^^^^^^^^^^^^^^^^
```
I checked the "dependencies" and "source-directories" listed in your elm.json,
but I cannot find it! Maybe it is a typo for one of these names?

from the elm compiler with:
`~/Documents/Development/SP-Responsive-Zoho/bdd/step_definitions$ npx elm-spec --specRoot ./`

I've done:
```
rm -rf elm-stuff/
rm elm.json.lock
```

and cleared cache:
`rm -rf ~/.elm`

elm --version
0.19.1

and tried re-starting VSCode etc.

I see that Spec.Step is exposed in the elm.json and it's listed in the docs.

Am I missing something? Should I be approaching this differently?

What do I need to do to import Spec.Step.Context as I believe I may need it to define the step?

Thanks in advance if you are able to assist ...

**Comment by brian-watkins** [comment]
Hello! A few things:

- `Spec.Step` is just a type alias for a function from a `Spec.Step.Context` to a `Spec.Step.Command`. 
- And `Spec.Step.Context` and `Spec.Step.Command` are both opaque types within the `Spec.Step` module, which means that the values they represent are used internally only.

But these are details that you usually don't need to worry too much about -- usually you won't need to write your own `Spec.Step` functions because you can just use the `Spec.Step`-generating functions that are already defined for you in elm-spec. So, for example, `Spec.Markup.Event.click` is a function that generates a Step for you (which means, it's a function that returns a function from a `Spec.Step.Context` to a `Spec.Step.Command`). So you can just use it directly when defining a Spec.

Your code sample indicates that you'd like to use a Step to "simulate the page loading". If by that you mean that you want to simulate the elm program starting then you don't need a separate step for that. Elm-Spec simulates the lifecycle of your program, which includes triggering the `init` function that is provided with `Spec.Setup.init` via the `given` function.

So, let's say that my program makes an HTTP request when it starts. I accomplish this by returning a `Cmd` value in my `init` function that makes the HTTP request:

```
init : Flags -> ( Model, Cmd Msg )
init flags =
  ( initialModel flags, myHttpCommand flags )
```

Now, when I write an elm-spec spec, I can provide this `init` function as part of the Setup (along with my `view` and `update` functions):

```
successStub =
  Spec.Http.Stub.for (Spec.Http.Route.get "http://fake-api.com/api/sports")
    |> Spec.Http.Stub.withText "something awesome"

Spec.describe "app start up"
[ Spec.scenario "the initial HTTP request succeeds" (
    Spec.given (
      Spec.Setup.init (init testFlags)
        |> Spec.Setup.withView view
        |> Spec.Setup.withUpdate update
        |> Spec.Http.Stub.serve [ successStub ]
    )
    |> Spec.it "displays the response from the server" (
      Spec.Markup.observeElement
        |> Spec.Markup.query << by [ id "stuff-from-server" ]
        |> Spec.expect (
          Spec.Claim.isSomethingWhere <|
          Spec.Markup.text <|
          Spec.Claim.isStringContaining "something awesome"
        )
    )
  )
]
```

When this Spec is executed, elm-spec will evaluate the `init` function to get the initial model and initial `Cmd msg`. It will then send that `Cmd msg` to be processed by the runtime just like it would be during normal program execution. So, if your `init` function returns a `Cmd msg` that triggers an HTTP request, this will happen automatically at the beginning of the Spec. Then you can use the `it` function to observe that the right thing happens.

Usually, if you are writing specs for code that triggers HTTP requests, you'll want to use functions in the `Spec.Http` module to stub responses to those requests. And then your spec can just observe that the right thing happens, given the stubbed responses that you provide in your spec. 

I'm happy to answer more questions if this isn't clear. Good luck!

**Comment by dotnetspec** [comment]
Hi there,

Thanks very much for your assistance.

I've made some adjustments based on your advice and the docs. To keep things as simple as possible initially I'm making a claim against an h1 tag for now. The Schedule page is now loading so there is some progress.

Currently, the test compiles but returns an error response:
`AvailabilityResponse (Err responseErr) ->`

after the http request in function 
Pages.Schedule.availabilityRequest :

```
initialModel : Pages.Schedule.FromMainToSchedule -> Model
initialModel flags =
    Pages.Schedule.initialize flags


init : Pages.Schedule.FromMainToSchedule -> ( Model, Cmd Msg )
init flags =
    ( initialModel flags, Pages.Schedule.availabilityRequest "" (initialModel flags) )


displayAvailableTimeslots : Spec Pages.Schedule.Model Pages.Schedule.Msg
displayAvailableTimeslots =
    let
        -- NOTE: placeholderURL is used to load the schedule page
        testFlags : Pages.Schedule.FromMainToSchedule
        testFlags =
            Pages.Schedule.FromMainToSchedule Nothing Consts.placeholderURL

        -- "https://squashpassion.com:443"
    in
    describe "Display Available Timeslots"
        [ scenario "User loads the schedule page, prompting a token response"
            (given
                -- Set up the initial conditions for the test
                (Spec.Setup.init
                    (init testFlags)
                    |> Spec.Setup.withView Pages.Schedule.view
                    |> Spec.Setup.withUpdate Pages.Schedule.update
                    |> Spec.Http.Stub.serve [ MySpecSetup.timeSlotsJsonStub ]
                )
                |> it "confirms page receives available timeslots"
                    (Markup.observeElement
                        |> Markup.query
                        --<< by [ id "timeSlot" ]
                        << by [ tag "h1" ]
                        |> Spec.expect
                            (Claim.isSomethingWhere <|
                                Markup.text <|
                                    --Claim.isStringContaining 1 "18:00"
                                    Claim.isStringContaining 1 "Availability Today"
                             
                            )
                    )
            )
        ]
```
In MySpecSetup module I have:
```
timeSlotsJsonStub : Spec.Http.Stub.HttpResponseStub
timeSlotsJsonStub =
    let
        jsonStr =
            "{\"data\": [ \"17:00\", \"18:00\", \"19:00\", \"20:00\", \"21:00\" ], \"timeZone\": \"Asia/Singapore\" }"

        jsonValue =
            case Json.Decode.decodeString Json.Decode.value jsonStr of
                Ok value ->
                    value

                Err err ->
                    --Debug.log "Invalid JSON: " err
                    Json.Encode.null

        responseBody =
            Spec.Http.Stub.withJson jsonValue
    in
    Spec.Http.Stub.for (Spec.Http.Route.get "http://localhost:5500/schedule")
       |> Spec.Http.Stub.withBody responseBody
```

so 
`Spec.Http.Stub.serve [ MySpecSetup.timeSlotsJsonStub ]`

doesn't have the intended effect (an Ok result for the availabilityRequest):
`AvailabilityResponse (Ok availableTimeSlotsJson) ->`

Please ignore my earlier attempts that may have been sent through to you. Apologies if I created any confusion.

What do I now need to do to obtain an Ok result for the availabilityRequest?

**Comment by dotnetspec** [comment]
I should include following with the above:

```
initialModel : Flags -> Model
initialModel flags =
    Pages.Schedule.initialize (Pages.Schedule.FromMainToSchedule Nothing Consts.placeholderURL)

init : Flags -> ( Model, Cmd Msg )
init flags =
    ( initialModel flags, Pages.Schedule.availabilityRequest ""  (initialModel flags))
```

**Comment by dotnetspec** [comment]
If you noticed, I pasted the wrong error initially. Apologies if that caused any inconvenience.

**Comment by dotnetspec** [comment]
It may help to know that my 'init' in Schedule.elm looks like:

```
init : FromMainToSchedule -> ( Model, Cmd Msg )
init fromMainToSchedule =
    ( 
      initialize fromMainToSchedule
    , Cmd.batch
        [ initialize fromMainToSchedule
            |> accessTokenRequest
        , getCurrentTimeTask
        ]
    )

type alias FromMainToSchedule =
    { time : Maybe DateTime
    , flagUrl : Url.Url
    }
```
As you probably realized, I need FromMainToSchedule mainly to send the current date/time from Main.elm in the full app.

**Comment by brian-watkins** [comment]
One minor thing first: when you're creating a Stub with a JSON body, it might be easier to use the `Json.Encode` functions to build up the response body. So in the case above you could write that as:

```
timeSlotsJsonStub =
  Spec.Http.Stub.for (Spec.Http.Route.get "http://localhost:5500/schedule")
    |> Spec.Http.Stub.withBody (Spec.Http.Stub.withJson <| 
      Json.Encode.object
        [ ("data", Json.Encode.list Json.Encode.string [ "17:00", "18:00", ... ])
        , ("timezone", Json.Encode.string "Asia/Singapore")
        ]
    )
```

That way you don't have to build a string and then decode it in your stub.

As for why your test isn't getting the (stubbed) response you expect ... one thing to try would be to print out (ie `Debug.log`) the response error you're getting. If it's a 404 then your app may be making a request to a URL that is not stubbed -- the Stub you provided will only match against requests sent to `http://localhost:5500/schedule` -- and I notice your Spec function mentions `https://squashpassion.com:443`. In any case, the error response might give a better indication of what the problem with the stub setup is ...

**Comment by dotnetspec** [comment]
Thanks, I updated the timeSlotsJsonStub function with an Elm encoded json body.

It's not a 404. I get an html response:
```
Compiling Elm ... Done!

Running specs: x (0.1s)

Accepted: 0
Rejected: 1

Failed to satisfy spec:

  /home/alanpoe/Documents/Development/SP-Responsive-Zoho/bdd/step_definitions/src/ScheduleAvailabilitySpec.elm

  Display Available Timeslots

  Scenario: User loads the schedule page, prompting a token response
    It confirms page receives available timeslots

    Claim rejected for selector
      h1

    Claim rejected for element text

    Expected
      Please submit the form below to book a timeslot at your convenience: 

    to contain 1 instance of
      Availability Today

    but the text was found 0 times
```
    
it's just that the response ('Please submit the form below to book a timeslot at your convenience') indicates that the json is deemed to have Errored by the app, which uses the AvailabilityResponse type constructor:
`type Msg = AvailabilityResponse (Result Http.Error Lvl1AvailResponse)`
Valid json, which the stub is intended to represent, should result in 'Availability Today' appearing in the h1 tag.

I have not been able to find a way to add Debug.log anywhere in the code e.g.:

```
timeSlotsJsonStub =
    let
        jsonObject =
            Json.Encode.object
                [ ( "response"
                  , Json.Encode.object
                        [ ( "returnvalue"
                          , Json.Encode.object
                                [ ( "reponse", Json.Encode.bool True )
                                , ( "data", Json.Encode.list Json.Encode.string [ "17:00", "18:00", "19:00", "20:00", "21:00" ] )
                                , ( "time_zone", Json.Encode.string "Asia/Singapore" )
                                ]
                          )
                        , ( "status", Json.Encode.string "success" )
                        ]
                  )
                ]
    in
    Spec.Http.Stub.for (Spec.Http.Route.get "http://localhost:5500/schedule")
        |> Spec.Http.Stub.withStatus 200
        |> Spec.Http.Stub.withBody
            (Spec.Http.Stub.withJson <| (Debug.log "jsonObject"  jsonObject))
```
without getting an
```
Compiling Elm ... Done!
    TypeError: this.log is not a function
```
error. What am I doing wrong?
 
(Incidentally, 'reponse' in the json is not a typo, the json is returned from the API like that.)

For ref the code that's triggered in update in the app from the availabilityRequest function that returns a Cmd Msg is:

```
    AvailabilityResponse (Ok availableTimeSlotsJson) ->
            let
                apiSpecs =
                    model.apiSpecifics

                -- NOTE: Decode the json response
                lvl1availResponse =
                    gotAvailResponse (Just availableTimeSlotsJson)

                newapiSpecs =
                    { apiSpecs | queryType = Avail lvl1availResponse }

                -- REVIEW: possible rf, to handle response data earlier (not in view)
                lslots =
                    case newapiSpecs.queryType of
                        Avail avail ->
                            case avail.response of
                                Just lvl1availResp ->
                                    lvl1availResp.response.returnValue.data

                                Nothing ->
                                    [ Nothing ]

                        _ ->
                            [ Nothing ]

                -- NOTE: Model updated to track query string changes
                newApiConfig =
                    --Zoho (Url Https Consts.zohoApisHost Nothing (Consts.zohoBookingsPath ++ "/availableslots") (Just (buildApiQueryString model)) Nothing)
                    Zoho (Url Https Consts.zohoApisHost Nothing Consts.zohoAvailableSlotsURL.path (Just (buildApiQueryString model)) Nothing)
            in
            ( { model
                | apiSpecifics = newapiSpecs
                , apiConfig = newApiConfig
                , status = Loaded
                , posixTmes =
                    convertTodaysAvailListToPosix (convertListMaybeToList lslots)
                        (case model.datetimeFromMain of
                            Just (CurrentDateTime dt _) ->
                                dt

                            Just (SelectedDateTime _ _) ->
                                DD.civilToPosix humandateTimePlaceholder

                            Nothing ->
                                DD.civilToPosix humandateTimePlaceholder
                        )
              }
            , Cmd.none
            )

       
        AvailabilityResponse (Err responseErr) ->
            let
                newModel =
                    { model | status = Errored, errors = model.errors ++ [ httpErrorToString responseErr ] }
            in
            ( newModel
            , Cmd.none
            )
```
My guess is that the issue relates to:
```
-- TYPE MISMATCH ------------------------------------------- src/MySpecSetup.elm

The 1st argument to `withJson` is not what I expect:

158|             (Spec.Http.Stub.withJson lvl1AvaildataResult)
                                          ^^^^^^^^^^^^^^^^^^^
This `lvl1AvaildataResult` value is a:

    Result Json.Decode.Error Lvl1AvailResponse

But `withJson` needs the 1st argument to be:

    Json.Encode.Value

Error running spec suite!
```

i.e. my app expects a Result, but the stub only returns a Json.Encode.Value. I can either give the stub what it requires (json value) or the app a Result, but I, currently, cannot see how to use them together (the stub returns a Result).

I see from the docs that I can observe a Result type ([observeResult](https://package.elm-lang.org/packages/brian-watkins/elm-spec/3.3.2/Spec-Observer#observeResult), but I can't see how to stub a Result (if I'm correct that that is the issue).

Does this mean there is a need for a new withResult stub within elm-spec? Or is there another way to resolve this with the existing functionality?

Thanks again ...

**Comment by dotnetspec** [comment]
Similar problem if provide withJson with a Json.Encode.Value and attempt to add the Result to Stub.serve:
```
The 1st argument to `serve` is not what I expect:

118|                     |> Spec.Http.Stub.serve [ Result.Ok MySpecSetup.timeSlotsJsonStub ]
                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
This argument is a list of type:

    List (Result error Spec.Http.Stub.HttpResponseStub)

But `serve` needs the 1st argument to be:

    List Spec.Http.Stub.HttpResponseStub
```

**Comment by brian-watkins** [comment]
For using `Debug.log` to help you see what's going on, I find it useful to add it to the update function so I can see every message. For example:

```
update : Msg -> Model -> ( Model, Cmd Msg)
update msg model =
  let
    d = Debug.log "msg" msg
  in
    case msg of
       ...
```

I put together a small repo with an example spec that describes the behavior of a program that sends an HTTP request on init. Hopefully that will help to show how all the pieces fit together.

https://github.com/brian-watkins/elm-spec-example

**Comment by dotnetspec** [comment]
The elm-spec-example helps a lot in clarifying how everything fits together when init triggers the first Cmd Msg in 
App.

My first Cmd Msg triggers `AccessTokenRequest` which triggers `AccessTokenResponse (Result Http.Error AccessTknResponse`)
I stub this response with:
```

accessTokenStub =
    let
        jsonObject =
            E.object
                [ ( "access_token", E.string "<access token>" )
                , ( "api_domain", E.string "https://www.zohoapis.com" )
                , ( "token_type", E.string "Bearer" )
                , ( "expires_in", E.int 3600 )
                ]
    in
    Stub.for (Route.get (testFlags.apiURL))
        |> Stub.withBody
            (Stub.withJson <|
                jsonObject
            )

```
using:
```
Spec.Setup.init
        (Pages.Schedule.init testFlags)
        |> Spec.Setup.withView Pages.Schedule.view
        |> Spec.Setup.withUpdate Pages.Schedule.update
        |> Spec.Http.Stub.serve [ MySpecSetup.accessTokenStub ]
```

and Debug.log in update informs me that accessTokenStub returns an Ok response.

My app only goes on to make an Availability request if I receive an Ok response from `AccessTokenResponse`. I am not sure if 
testing `AvailabilityResponse` is dependent on simulating the receipt of an `Ok AccessTokenResponse `or not(?). In either case I do not know how I can simulate the Availability request and utilize the `timeSlotsJsonStub`.

Is it necessary for me to go via:
```
Spec.Setup.init
        (Pages.Schedule.init testFlags) etc...
```
for all the functionality after an `Ok AccessTokenResponse` and if so how do I move to the next step (after `|> Spec.Http.Stub.serve [ MySpecSetup.accessTokenStub ]`), namely
receiving an `AvailabilityResponse (Result Http.Error Lvl1AvailResponse)`?

Thanks again ...

**Comment by brian-watkins** [comment]
elm-spec will continue to simulate the commands in your program until there are no more (ie, until your update function returns a `Cmd.none`). So, if your program makes multiple HTTP requests all you need to do is provide appropriate stubs for those HTTP requests -- that's why `Spec.Http.Stub.serve` takes a list of `Spec.Http.Stub.HttpResponseStub` values.

If I understand the case you're describing correctly, you could update your spec setup like so:

```
Spec.Setup.init
        (Pages.Schedule.init testFlags)
        |> Spec.Setup.withView Pages.Schedule.view
        |> Spec.Setup.withUpdate Pages.Schedule.update
        |> Spec.Http.Stub.serve [ MySpecSetup.accessTokenStub, timeSlotsJsonStub ]
```

I've updated the example repo to describe something like the behavior I think you're after -- the example program now makes one HTTP request for an access token on init and, if that is successful, it makes another request for data to display.

https://github.com/brian-watkins/elm-spec-example

**Comment by dotnetspec** [comment]
There is progress as I'm able to run your example successfully and my own code now mirrors the pattern and compiles.

However, I'm currently unable to pass the last 2 of the 3 tests you specified, as I cannot yet obtain an `AvailabilityResponse Ok `and the relevant timeslots. 
I receive a `404 (Not found) `error.

`AvailabilityResponse `is triggered by a call to `accessTokenRequest`(working).
My code is:
```

AvailabilityResponse (Result Http.Error Lvl1AvailResponse)

availabilityRequest : String -> Model -> Cmd Msg
availabilityRequest serviceId model =
    let
        proxyParams =
            Maybe.withDefault (ToProxyConfig "" [] "testPostRequest error" Http.emptyBody Nothing Nothing) model.toProxyConfig

        therequest =
            Http.request
                { body = Http.emptyBody
                , method = proxyParams.method
                , headers = proxyParams.headers

                -- NOTE: serviceId is updated for each venue
                , url = proxyParams.url ++ buildApiQueryString (updateServiceId serviceId model)
                , expect = Http.expectJson AvailabilityResponse lvl1AvaildataDecoder
                , timeout = Nothing
                , tracker = Nothing
                }
    in
    therequest

AvailabilityResponse (Ok availableTimeSlotsJson) ->
            (model updates
            , Cmd.none
            )


AvailabilityResponse (Err responseErr) ->
    ( model updates
    , Cmd.none
    )

successfulThingsStub =
    Stub.for (Route.get "http://localhost/proxy?apiUrl=https://www.zohoapis.com/bookings/v1/json/availableslots&query_type=bookingsAvailability&service_id=4503471000000091024&staff_id=4503471000000033016&selected_date=01-Jan-1970%208:00:00")
        |> Stub.withBody
            (Stub.withJson <|
                Consts.validWithValuesJsonResponeObj
            )

validWithValuesJsonResponeObj =
    E.object
        [ ( "response"
          , E.object
                [ ( "returnvalue"
                  , E.object
                        [ ( "reponse", E.bool True )
                        , ( "data", E.list E.string [ "17:00", "18:00", "19:00", "20:00", "21:00" ] )
                        , ( "time_zone", E.string "Asia/Singapore" )
                        ]
                  )
                , ( "status", E.string "success" )
                ]
          )
        ]
```
I believe that `validWithValuesJsonResponeObj` is valid json and that it matches the relevant decoder because the same fragment passes the unit tests with it.

Your code, for comparison, is:

```
FetchedThings (Result Http.Error (List String))

fetchThings : String -> AccessToken -> Cmd Msg
fetchThings baseUrl accessToken =
  Api.authenticatedRequest
    { accessToken = accessToken
    , url = baseUrl ++ "/things"
    , expect = Http.expectJson FetchedThings (Json.list Json.string)
    }

FetchedThings result ->
      case result of
        Ok things ->
          ( { model | state = State.withThings model.state things }, Cmd.none )
        Err _ ->
          ( { model | state = State.withThings model.state State.defaultThings }, Cmd.none )

successfulThingsStub =
  Stub.for (Route.get testThingsUrl)
    |> Stub.withBody (Stub.withJson <|
      Encode.list Encode.string [ "birds", "fish", "trees" ]
    )
```


The relevant spec is:

```
mangageTokenAndDisplayAvailableTimeslots : Spec Sched.Model Sched.Msg
mangageTokenAndDisplayAvailableTimeslots =
    describe
        "An authenticated HTTP request is made to fetch things when the app starts"
        [scenario "both requests are successful"
            (given
                (Setup.init (Sched.init testFlags)
                    |> Setup.withUpdate Sched.update
                    |> Setup.withView Sched.view
                    |> Stub.serve
                        [ successfulAccessTokenStub
                          ,
                          successfulThingsStub
                      
                        ]
                )
                |> Spec.observeThat
                    [ it "sends the access token in the request for things"
                        (Spec.Http.observeRequests (Route.get "http://localhost/proxy?apiUrl=https://www.zohoapis.com/bookings/v1/json/availableslots&query_type=bookingsAvailability&service_id=4503471000000091024&staff_id=4503471000000033016&selected_date=01-Jan-1970%208:00:00")
                            |> expect
                                (Claim.isListWhere
                                    [ Spec.Http.header "Authorization" <|
                                        Claim.isSomethingWhere <|
                                            Claim.isStringContaining 1 "Zoho-oauthtoken 1000.30251938ab93b2cf5088c2d9fee4926d.4cb1807a8ef0f5e7c30e89dd144683e2"
                                    ]
                                )
                        )
                    ]
            )
        ]
```
I have tried:
```
|> Stub.serve
                        [ successfulAccessTokenStub
                          ,
                          successfulThingsStub
                          ]
```
the other way round: 

```
|> Stub.serve
                        [ successfulThingsStub 
                          ,
                          successfulAccessTokenStub
                          ]
```
 as well.    
                          
With this setup, I get the following Debug msgs in AvailabilityResponse (Err responseErr):

```
toProxyConfig: Just { body = <internals>, headers = [Header "Authorization" "Zoho-oauthtoken 1000.30251938ab93b2cf5088c2d9fee4926d.4cb1807a8ef0f5e7c30e89dd144683e2"], method = "GET", timeout = Nothing, tracker = Nothing, url = "http://localhost:3000/proxy" }

datetime: "01-Jan-1970 8:00:00"

model apiSpecifics: { accessToken = Just "1000.30251938ab93b2cf5088c2d9fee4926d.4cb1807a8ef0f5e7c30e89dd144683e2", maxResults = "", 

queryType = Avail { response = Nothing, service_id = "4503471000000091024", slots = [], staff_id = "4503471000000033016" } }

```
And this msg:
`msg: AvailabilityResponse (Err (BadStatus 404))`

The route in `successfulThingsStub` is the same as the route in `observeRequests` (please see attachment):
![SP - Availability text compare - Screenshot from 2024-03-01 09-32-34](https://github.com/brian-watkins/elm-spec/assets/27340988/563df27e-e118-40d5-9277-f1bc8b347d6f)

The test output is:
```
Accepted: 1 (the AccessTokenResponse)
Rejected: 1 (the availabilityRequest header)

Failed to satisfy spec:

  /home/alanpoe/Documents/Development/SP-Responsive-Zoho/bdd/step_definitions/src/ScheduleAvailabilitySpec.elm

  An authenticated HTTP request is made to fetch things when the app starts

  Scenario: both requests are successful
    It sends the access token in the request for things

    Claim rejected for route
      GET http://localhost/proxy?apiUrl=https://www.zohoapis.com/bookings/v1/json/availableslots&query_type=bookingsAvailability&service_id=4503471000000091024&staff_id=4503471000000033016&selected_date=01-Jan-1970%208:00:00

    List failed to match

    Expected list to have length
      1

    but it has length
      0
```

I'm not sure if the header test is dependent on a successful availability response, but it appears to be.

Do you know why I might be getting a 404 Not Found here?

Is there anything else I might have missed?

Thanks again ...

**Comment by brian-watkins** [comment]
I believe you would get a `404` response code if your program attempts to make a request that is not stubbed. One thing that could help you debug what's going on is to use [Spec.Http.logRequests](https://package.elm-lang.org/packages/brian-watkins/elm-spec/latest/Spec-Http#logRequests). This is a step you can use to log any requests that have been received so far. So, in your spec, add something like:

```
scenario "both requests are successful" 
    (Spec.given
        (Setup.init ...)
    )
    |> Spec.when "debug after initial command is processed"
        [ Spec.Http.logRequests
        ]
   |> Spec.observeThat ...
```

And this should print out on the console all the requests that your program is making. Maybe that will help you see what your program is trying to do so that you can adjust your stubs accordingly. Good luck!

**Comment by dotnetspec** [comment]
Yes, that's resolved it.

I'm now stubbing token and timeslot requests with json responses and matching them with the display fields in the UI.

My app makes multiple (5) separate requests with different service_ids. I had to ensure the request url I was observing in the scenario matched the _last_ request in the list and that that request url is the one that I stubbed. The first 4 would, I presume, have resulted in 404 errors, but they could be ignored.

Thanks again for your patience and helpful pointers ...
